### ISSUE TITLE
Multiple languages, multiple toolsets, many problems

### ISSUE THREAD

**Issue by TurkeyMan** [premake/premake-core#1215]
So this is just a space for discussion... since I'm not really sure how I should approach this problem space.

Premake has `language`, and `toolset`, and they are singular values. I'm working with multi-language projects more and more, and we don't seem to be able to express this well...

In one case, I had an Android project where I wanted to have C++, D, and Java code all together in the one makefile. I started trying to hack premake to see if I could support this configuration, but realised we have some expressive limitations in the way.

Firstly, what `language` is this project? I hacked this multi-language support into `language "C++"`, but that's a big red-flag right off the bat.
The next hurdle is that each language has their own sense of toolset.
C++ code wants `toolset "clang"`, D code wants `toolset "ldc"`, and Java needs to have a toolset specified too...

The next issue I ran into was that I didn't want to pollute the makefile with configuration block variables for the D or Java toolsets eagerly, I'd rather see if there are actually any `.d` or `.java` files in the project before I worry about emitting build environment configuration for those languages. This same issue would apply to C++ too.

Basically, I think it looks something like this:

- There are many languages that exist.
- They all define file patterns for their respective source files.
- We need to know post-oven which languages are actually present in the project (also per configuration?), so the backend can emit the right stuff
- Languages each have their own independent `toolset` setting
- Specifying a single `language` for the project seems wrong

So... is this a filtration problem? That doesn't feel right. Do we filter for `language:X` and then assign `toolset` against that filter?
How should we describe how to detect if a language has any presence in a project such that we can query for that presence in the backend? Are languages registered in some way with some file patterns?

I'd love to hear peoples independent design thoughts. I think this might lead to a fairly aggressive structural change, and I'm not sure how to just 'patch it in there' nicely. I think we got some core concepts wrong. Projects don't have one `language`, and one `toolset`, they may have many.

**Comment by tritao** [comment]
Hey, @TurkeyMan.

This is an interesting problem, and it also goes inline with some other uses cases I've been thinking, specifically, having a WebAssembly target that can combine modules from multiple languages in the same project.

I think trying to shoehorn multiple language support in the existing `project` abstraction might lead to weird conceptual problems like the ones you mentioned.

Projects already know how to compile their source files into outputs, so we could leverage that and combine them in top-level projects. These top-level projects, could have a different terminology (`target` comes to mind) or just be a variant of the existing `project`,  and would know how to combine the outputs of dependent projects in a final target.

In the case of Android this would probably have its own specific backend, which could take both Java/Kotlin (DEX bytecode) and native (C++, D) projects (object files).

In the case of WebAssembly, it could also take projects of multiple languages, as long as their backends know how to output LLVM bitcode. 

```lua
project "Core"
    language "C"
    toolset "emscripten"
    kind "StaticLib"

project "LibA"
    language "TS"
    typescript "tsc-wasm"
    kind "StaticLib"

project "LibB"
    language "C#"
    toolset "mono-wasm"
    kind "StaticLib"

target "WebApp"
    kind "WebApp"
    links "Core", "LibA", "LibB"
```

This assumes that static libraries can be created for all languages we want to support. This seems like a reasonable assumption to make in this case, but didn't put much thought into it. The target would make sure the dependencies are up-to-date and call the linker to combine all the outputs from the dependent projects.

Now that I think about it, I am not sure if we even need a different abstraction. `target` would just be a `project` but with limitations around most of the API which would not make sense.

This also brings some other questions to me, which is how we should deal with generating these multiple language projects to users / IDEs. 

Should we generate all the files from multiple languages in the same project? If that's an expectation then this design would probably not have the best terminology from end-user point.

But I'd say it's easier to keep them in different projects, so we can re-use most of the existing abstractions and not complicate things too much.

Happy to hear other thoughts about this as well.

**Comment by TurkeyMan** [comment]
> Should we generate all the files from multiple languages in the same project?

Yes, and that's where my struggle lies

> But I'd say it's easier to keep them in different projects

That's not what I'm talking about, and not what I want

I've been feeling for quite some time that a lot of IDE's should just wrap around a makefile (or ninja ideally).
XCode can *barely* express premake projects, and it would be better just to wire the build targets up to make/ninja targets, but take care to populate the project metadata such that debugging and autocomplete work properly. There's no point trying to continue to fight against XCode build system.
Same story for some other IDE's.
Then there's IDE's like VSCode which don't have a build system at all, which really just need to wrap some project description around make/ninja again. I think we should start to focus on that for niche IDE's.

But that's all an aside... the effective structure is this:
We define build rules for multiple languages, and include files built by those rules together in one project.
This works somewhat with make (gmake2 is all `rule` driven, even for C/C++), and VS might ideally be improved to work this way too.
Perhaps if C/C++ wasn't special, and instead was implemented in terms of a normal language `rule`, then we'd come to understand the deficiencies in rule definitions limiting other languages.
Expressing a `rule` in make, or in MSBuild, or in ninja should all probably be core features of premake, then language definitions kinda just come for free.

**Comment by tritao** [comment]
So let me see if I understand you correctly: your end goal is to get all the different languages in the same Makefile/ninja generated file?

Why does it matter if all the rules for the multiple language project are in the same file or separate in different `<project>.make> files like we generate right now in Premake? The end result should be the same independently if they're on the same Makefile or in multiple one-per-project Makefiles, right?

Exactly to allow everything to work correctly from an IDE standpoint (debugging, code complete) is why I suggested generation of multiple different projects per language, and just link them together in the build. Though it depends on the IDE of course, for VS Code it doesn't matter as you said.

Agree with you about ninja, I've been using it more and more and would be awesome to get this as core feature of Premake. If I understand your suggestion correctly, you think language support could be implemented generically as part of a core rule abstraction, and then could be shared between Make/ninja/MSBuild. That's a great idea, and could lead to less code duplication / feature parity between backends.

**Comment by tvandijck** [comment]
I think you put the nail on the head with the `rule` based approach. When I wrote the `gmake2` backend I sort of realized that that was ultimately the abstraction I wanted to center around for everything. And to some extent I got there with that backend, but it sort of bolted the C/C++ and Linker rules in there in a hard-coded fashion.

I also think the `rule` keyword isn't done as an API... For example it defines an executable to run as part of the rule, but that should really be the task of the `toolset`, and not the rule. The rule should simply have a `toolset` setting, with a default setting.

But there is many more things if you go that path. *Any* setting currently in premake should basically become a setting of a particular `rule` and the `toolset` should be able to translate those settings into command line arguments. For example the `disablewarnings` is currently a global API, but we then have a bunch of special code in the backends to 'tie' that to the C++/C rule really... and that is true for basically *all* API's, they basically map to a specific part of the pipeline, some apply to the compiler 'rule' and some apply to the linker 'rule'.

In my opinion, what I would like to write is:

```lua
project 'foobar'
    rules { 'cpp', 'D', 'Java' }

    sourcefiles { '*.cpp', '*.d', '*.java' }
    nonsourcefiles { '*,h', '*.inl' }

    cppRuleVars {
        toolset = 'clang',
        disablewarnings = { 4100, 4250 }
    }

    dRuleVars { 
        ...
    }
 
etc..
```

Where this sort of becomes 'ugly', or less then ideal is the way the rule variables are now specified. They are not longer function calls, but it all goes into this somewhat unstructured table, and the syntax of that is far from ideal. So that is the part that needs solving.

But I do feel that this would be a much more powerful abstraction and would map to makefiles, ninja and a bunch of other make systems a lot more naturally. It will be harder to map to a visual studio project maybe.

**Comment by tvandijck** [comment]
come to think of what I wrote above.. maybe even the `sourcefiles` API should be defined by the rule, since different rules require different sets of files...

- C# has sourcefiles, resourcefiles, and a bunch of others.
- C/C++ has sourcefiles and headerfiles.
- etc.

**Comment by samsinsane** [comment]
Regarding multiple languages, I'm significantly more interested in Obj-C/C++ and Swift working with Xcode projects and Makefiles. If getting that working happens to assist with Android+D, great but otherwise, I'm not overly in favour of a rewrite of chunks of Premake just for that. I just find it hard to justify given how small the group of users are that it would benefit.

**Comment by TurkeyMan** [comment]
> Where this sort of becomes 'ugly', or less then ideal

Is it a job for `filter`?

```lua
filter { "rule:cpp" )
  toolset 'clang'
filter { "rule:linker" )
  outdir "some/where"
```

??

**Comment by TurkeyMan** [comment]
> maybe even the sourcefiles API should be defined by the rule [...]

I kinda feel like maybe not, and let the rules describe patterns to select files... otherwise you can't use things like `src/**` anymore, which might contain bunches of different code.

**Comment by tvandijck** [comment]
> Is it a job for filter?

It could be I guess, but I would like to see an error if a particular API doesn’t apply to a rule. For example setting the ‘vectorextensions’ in a ‘filter { “rules:linker” }’ probably makes no sense. So it would be helpful if premake said something.

**Comment by TurkeyMan** [comment]
Right... but this is a huge development, and a gigantic departure from premake 5.x... it's practically a different frontend at this stage :/

**Comment by TurkeyMan** [comment]
API's would need to be registered against rules, rather than globally.
I guess the API registration info table can have a rule field added...

**Comment by tvandijck** [comment]
>  and a gigantic departure from premake 5.x

Yup, can barely call is 5.x at that point, or even ‘premake’. The only thing left standing would be the fact that is uses lua.

**Comment by TurkeyMan** [comment]
Pretty much... which leads to deep questions ;)

**Comment by TurkeyMan** [comment]
I know... it quite upsets me that usage thing has stalled... because we still maintain the package thing here, which is awkward and most of its problem space would be resolved with your usage PR... our internal package manager could just be a small extension that manages fetching packages.

**Comment by TurkeyMan** [comment]
Yeah... between all these things, maybe we should head back to the drawing board on frontend design :/
But that's probably terminal for premake in general. Everything's hard!

**Comment by starkos** [comment]
I might have an opportunity to revisit some of these major architectural roadblocks. Putting aside backwards-compatibility for now, what *should* this feature look like?

I think we're going to have to draw up two paths: how it should work in "restricted" project formats like VC++, and "free-form" formats like makefiles. Ideally we'd provide an API that let's you run wild in makefiles, while still providing a useful set of functionality that works across the board.

For starters, we absolutely will need to…

- Break up the current monolithic toolset APIs to allow compilers, linkers, and other tools to be mixed-and-matched
- Get rid of the project-level `language`

In theory, we could specify per-language settings using filters:

```lua
filter { 'files:**.h or **.cpp' }
    toolset 'clang'
    disableWarnings { 4100, 4250 }
```

That could (in a perfect future world) be wrapped up in a [reuse block](https://github.com/premake/premake-core/issues/1346) to simulate the rules:

```lua
block 'cpp'
    filter { 'files:**.h or **.cpp' }
        toolset 'clang'
        disableWarnings { 4100, 4250 }

project 'MyProject'
    uses { 'cpp' }
```

Premake would have to be smart enough to recognize that the filter applies to the entire project and not spam the generated project with a bunch of duplicate configuration.

Does that bring things any closer?
